File: c++/design_pattern/Mediator.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;


namespace BasicMediator {
	
	struct Mediator;

	struct ComponentA{

		ComponentA() : mediator(nullptr){}

		void opeationA();
		void set_mediator(Mediator* mediator){
			this->mediator = mediator;
		}
		
		private : 
			Mediator* mediator;
	};

	struct ComponentB{

		ComponentB() : mediator(nullptr){}

		void opeationB();

		void set_mediator(Mediator* mediator){
			this->mediator = mediator;
		}

		private :
			Mediator* mediator;
	};


	struct Mediator{

		ComponentA* a;
		ComponentB* b;

		Mediator(ComponentA* a, ComponentB* b) : a(a), b(b){
			a->set_mediator(this);
			b->set_mediator(this);
		}

		void notify(const string& sender, const string& recevier){
			if(sender == "A")
				cout << "Mediator react to A and  Send Message to " << recevier << endl;

			else if(sender == "B")
				cout << "Mediator react to B and  Send Message to " << recevier << endl;

			
		}
	};

	void ComponentA::opeationA(){
		cout << "A does operation" << endl;
		mediator->notify("A", "B");
	};
	void ComponentB::opeationB(){
		cout << "B does operation" << endl;
		mediator->notify("B", "A");
	};

	void main(void){
		cout <<"[BasicMediator]" << endl;

		ComponentA a;
		ComponentB b;
		Mediator m(&a, &b);

		a.opeationA();
		cout << endl;
		b.opeationB();
		cout << endl;

	}
}

namespace CollectionMediator {


	struct ChatRoom;

	struct Person{
		string name;
		ChatRoom* room = nullptr;
		vector<string> chat_log;

		Person(const string& name) : name(name){}

		void say(const string& message)const;
		void pm(const string& who, const string& message) const;
		void recieve(const string& origin, const string& message) {
			string s{origin+" : "+message+"\t"};
			cout << "[" << name<<"'s session]  " << s << endl;
			chat_log.emplace_back(s);
		}
	};

	struct ChatRoom{
		vector<Person*> people;

		void join(Person* p){
			string join_msg = p->name + " joins the chat";
			p->room = this;
			broadcast("rooms", join_msg);
			people.push_back(p);
		}

		void broadcast(const string& origin, const string& message){
			for(auto p : people)
				if(p->name != origin)
					p->recieve(origin, message);
			
		}

		void message(const string& origin, const string& who, const string& message){
			auto target = find_if(begin(people), end(people),[&](const Person* p){
				return p->name == who;
			});

			if(target != end(people))
				(*target)->recieve(origin, message);
		}
	};

	void Person::say(const string& message)const {
		room->broadcast(name, message);
	}
	void Person::pm(const string& who, const string& message) const {
		room->message(name,who,message);
	}



	void main(void){

		cout <<"[CollectionMediator]" << endl;

		ChatRoom room;

		Person john{"john"};
		Person jane{"jane"};
		Person simon{"simon"};

		room.join(&john);
		room.join(&jane);
		john.say("hi jane");
		room.join(&simon);
		john.say("hi simon");
		john.pm("simon", "glad  you could join us, simon!");



		cout << endl;
	}
}


namespace MediatorWithHierahicalComponent {

	class BaseComponent;
	class Mediator{
		public :
			virtual void notify(BaseComponent* sender, const string& event) const = 0;
	};

	class BaseComponent {
		protected : 
			Mediator* mediator;

		public :
			string name;
			BaseComponent() : mediator(nullptr){}
			BaseComponent(const string& name) : mediator(nullptr), name(name){}
			BaseComponent(const string& name, Mediator* mediator) :name(name), mediator(mediator){}
			
			virtual void set_mediator(Mediator* mediator){
				this->mediator = mediator;
			}
	};

	struct ComponentA : public BaseComponent{

		explicit ComponentA(const string& name) : BaseComponent(name){}

		void doA(){
			cout << "Component " <<  name << " does A" << endl;
			mediator->notify(this, "A");
		}

		void doB(){
			cout << "Component " <<  name << " does B" << endl;
			mediator->notify(this, "B");
		}

	};

	struct ComponentB : public BaseComponent{


		explicit ComponentB(const string& name) : BaseComponent(name){}

		void doC(){
			cout << "Component " <<  name << " does C" << endl;
			mediator->notify(this, "C");
		}

		void doD(){
			cout << "Component " <<  name << " does D" << endl;
			mediator->notify(this, "D");
		}

	};

	class ConcreteMediator : public Mediator {
		ComponentA* a;
		ComponentB* b;

		public :

			explicit ConcreteMediator(ComponentA* a, ComponentB* b) : a(a), b(b){
				this->a->set_mediator(this);
				this->b->set_mediator(this);
			}


			void notify(BaseComponent* sender, const string& event)const override{

				if(auto child = dynamic_cast<ComponentA*>(sender))
					if(event == "A"){
						cout << "Mediator reacts "<<child->name <<" on A and trigger following operations:" << endl;
						this->b->doC();
					}
				if(auto child =dynamic_cast<ComponentB*>(sender))
					if( event == "D"){
						cout << "Mediator reacts "<<child->name << " on B and trigger following operations:" << endl;
						this->a->doB();
					}
			}
	};


	void main(void){
		cout <<"[MediatorWithHierahicalComponent]" << endl;
		ComponentA* a = new ComponentA("<a1>");
		ComponentB* b = new ComponentB("<b1>");

		ConcreteMediator* mediator = new ConcreteMediator(a,b);

		a->doA();
		cout << endl;
		b->doD();

		delete a;
		delete b;
		delete mediator;

	}
}



namespace MediatorWithBoostSignal {
	void main(void){
		cout <<"[MediatorWithBoostSignal]" << endl;
		cout << endl;

	}
};


int main(void){
	BasicMediator::main();
	cout << endl;
	CollectionMediator::main();
	cout << endl;
	MediatorWithHierahicalComponent::main();
	cout << endl;
	MediatorWithBoostSignal::main();
	cout << endl;

	return 0;
}

File: java/src/main/java/com/example/design_pattern/Mediator.java
package com.example.design_pattern;

import java.util.*;

public class Mediator {

  private static class BasicMediator {

    private static class ComponenetA {
      private Mediator_ mediator;

      public void setMeidator(Mediator_ mediator) {
        this.mediator = mediator;
      }

      public void operationA() {
        System.out.println("A does OperationA");
        mediator.notify("A", "B");
      }

      public void receive(String message) {
        System.out.println("A recive message: " + message);
      }
    }

    private static class ComponentB {
      private Mediator_ mediator;

      public void setMeidator(Mediator_ mediator) {
        this.mediator = mediator;
      }

      public void operationB() {
        System.out.println("B does OperationB");
        mediator.notify("B", "A");
      }

      public void receive(String message) {
        System.out.println("B recive message: " + message);
      }
    }

    private static class Mediator_ {

      private ComponenetA a;
      private ComponentB b;

      Mediator_(ComponenetA a, ComponentB b) {
        this.a = a;
        this.b = b;
        a.setMeidator(this);
        b.setMeidator(this);
      }

      public void notify(String sender, String reciever) {
        if ("A".equals(sender)) {
          System.out.println(String.format("Mediator react to A and Send Message to %s", reciever));
          a.receive("Hello World!");
        }

        if ("B".equals(sender)) {
          System.out.println(String.format("Mediator react to B and Send Message to %s", reciever));
          b.receive("Hello World!");
        }
      }
    }

    public static void main() {
      ComponenetA a = new ComponenetA();
      ComponentB b = new ComponentB();

      Mediator_ mediator = new Mediator_(a, b);
      a.operationA();
      b.operationB();
    }
  }

  private static class CollectionMediator {

    private static class Person {
      public String name;
      public ChatRoom chatroom;
      public List<String> chatLogs = new ArrayList<>();

      Person(String name) {
        this.name = name;
      }

      public void say(String message) {
        if (chatroom != null) {
          chatroom.boradcast(name, message);
        }
      }

      public void privateMessageTo(String whom, String message) {
        if (chatroom != null) {
          chatroom.message(name, whom, message);
        }
      }

      public void receiveFrom(String whom, String message) {
        System.out.println(String.format("[%s's session] %s : %s", name, whom, message));
      }
    }

    private static class ChatRoom {

      List<Person> people = new ArrayList<>();

      public void join(Person p) {
        String joinMsg = String.format("%s joins the chat", p.name);
        p.chatroom = this;
        boradcast("rooms", joinMsg);
        people.add(p);
      }

      public void boradcast(String from, String message) {

        for (Person p : people) {
          if (!from.equals(p.name)) {
            p.receiveFrom(from, message);
          }
        }
      }

      public void message(String from, String to, String message) {
        Optional<Person> optPerson = people.stream().filter(p -> to.equals(p.name)).findFirst();

        if (!optPerson.isEmpty()) {
          optPerson.get().receiveFrom(from, message);
        }
      }
    }

    public static void main() {

      Person john = new Person("John");
      Person jane = new Person("Jane");
      Person simon = new Person("Simon");

      ChatRoom chatroom = new ChatRoom();

      chatroom.join(john);
      chatroom.join(jane);
      john.say("hi Jane!");
      chatroom.join(simon);
      john.say("hi Simon!");
      john.privateMessageTo("Simon", "glad you could join us, Simon!");
    }
  }

  private static class MediatorWithHierarchicalComponent {

    private static interface Mediator_ {
      void notify(BaseComponent sender, String event);
    }

    private static class BaseComponent {
      protected Mediator_ mediator;

      public void setMeidator(Mediator_ mediator) {
        this.mediator = mediator;
      }
    }

    private static class ComponentA extends BaseComponent {
      void doA() {
        System.out.println("ComponentA does A");
        if (mediator != null) mediator.notify(this, "A");
      }

      void doB() {
        System.out.println("ComponentA does B");
      }
    }

    private static class ComponentB extends BaseComponent {
      void doC() {
        System.out.println("ComponentA does C");
      }

      void doD() {
        System.out.println("ComponentB does D");
        if (mediator != null) mediator.notify(this, "D");
      }
    }

    private static class ConcreterMediator implements Mediator_ {

      ComponentA a;
      ComponentB b;

      ConcreterMediator(ComponentA a, ComponentB b) {
        this.a = a;
        this.b = b;
        this.a.setMeidator(this);
        this.b.setMeidator(this);
      }

      @Override
      public void notify(BaseComponent sender, String event) {
        if (sender instanceof ComponentA) {
          if ("A".equals(event)) {
            System.out.println("Mediator react to A and trigger following operations: ");
            this.b.doC();
          }
        }
        if (sender instanceof ComponentB) {
          if ("D".equals(event)) {
            System.out.println("Mediator react to B and trigger following operations: ");
            this.a.doB();
          }
        }
      }
    }

    public static void main() {

      ComponentA a = new ComponentA();
      ComponentB b = new ComponentB();

      ConcreterMediator mediator = new ConcreterMediator(a, b);

      a.doA();
      b.doD();
    }
  }

  public static void main(String... args) {
    BasicMediator.main();
    CollectionMediator.main();
    MediatorWithHierarchicalComponent.main();
  }
}
